### PR TITLE
Feat/bulk/all lambda

### PR DESCRIPTION
**Pull request by jimvae**
## Approach
The Lambda is triggered via SQS, and takes in the following in the message body (below shown in JSON format for readability but input message body should be JSON-stringified):
```
{
  filePath: "aopmo234o_1", // folder path in S3 bucket where files are stored
  mappings: [
    {
      shortUrl: "abc_0",
      longUrl: "https://google.com"
    },
    {
      shortUrl: "abc_1",
      longUrl: "https://google.com"
    },
  ]
}
```
The CSV file is generated and uploaded directly to S3. QR codes are generated and saved in Lambda ephemeral storage. After generating QR codes, a zip stream is opened to stream the file directory directly into a zip file on S3. Upon completion, we send a success or failure message to an SNS topic. See below for more details.
 
### Saving in /tmp directory
QR codes are generated and saved to `/tmp` directory. On average, 1000 png files takes up 43MB of space, and 1000svg files takes up 12MB of space. Since the `/tmp` ephemeral storage on the Lambda has 512 MB of space, this is no issue. 

We remove the created directories after the function has completed or upon catching any error for idempotency (no error should be thrown if the function is rerun). 

### Streaming Zip to S3
We use `archiver` to zip and stream the file to S3 directly. Zipping the directory within the Lambda would load all the files into buffer memory, so instead we create a zip stream and stream into S3 directly.

### SQS vs SNS
After experimenting a little, we have decided to use an SNS topic to receive the success/failure messages, as configuring the SQS polling mechanism on the backend is more tedious and did not work well with localstack. As we want to receive success/failure events in real time, SNS also fits the use case better.

## Env variables
`ASSET_VARIANT` -> gov, edu or health
`BULK_GENERATION_BUCKET` -> s3 bucket name to upload to
`DOMAIN` -> domain url of QR Code
`SNS_TOPIC_ARN` -> sns topic to publish success or failure messages
`SQS_END_QUEUE_URL` -> sqs queue to receive success or failure messages (unused currently as we're using SQS instead)

## Testing
To test out the function, go to the AWS Lambda console, click on `health-staging-bulk-qrcode-generation` function (the other staging functions are not set up yet). 

Go to `Test` tab, and click on `Edit saved event`. There should be a shared test event named `test_100_mappings`. See below.
Screenshot 2022-11-08 at 17 38 57

The test body should be populated for you. Feel free to modify `filePath`, under the `body` value - this will be the file directory path in S3 where your files will be saved. 

Press `Test` button to run the test. If the function is successful, the invocation bar should turn green, and the S3 bucket `health-bulk-qrcode-generation` should contain your files, saved under the `filePath` value that you provided.

### Manual tests
- [x] Should be able to run function consecutively with same input without error (idempotent)
- [x] Download SVG files and verify that the number of files are correct
- [x] Downloaded SVG QRCodes can be scanned and redirects to the correct url
- [x] Download PNG files and verify that the number of files are correct
- [x] Downloaded PNG QRCodes can be scanned and redirects to the correct url
- [x] Downloaded CSV should contain correct mappings
- [x] Success/failure message should send to SNS and be received

## Setup
After initial Serverless deployment, a few manual steps are required:
- Adjust timeout to 2mins (usually takes around 46 seconds for 1000 files)
- Add env variables 
- Add trigger
- Modify lambda role to allow SQS read, S3 write, SNS publish permissions (might not be necessary if set up already)

### CODE REVIEW

**Comment by gweiying** [comment]
please refer to comments left on #2032

**Review by halfwhole — APPROVED** [review]
LGTM, tested it out and the downloaded QR codes look good, really appreciate the detailed notes esp on testing :D unfortunately I don't know enough of the bigger picture (what happens before SQS -> lambda, and what happens after lambda -> SNS), but as this stands by itself it looks great!

some other thoughts:
- what access controls would we like for the S3 bucket for QR codes?
- would it be good to migrate our lambda functions from javascript -> typescript eventually?
- will we be using `sqs.js` in the future?
- seems to be some code duplication w the main codebase for qr code generation, but indeed there doesn't seem to be much we can do about it :')

File: src/server/serverless/bulk-qrcode-generation/index.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    await s3Service.uploadToS3(
+      csvBuffer,
+      'text/csv',
+      `${filePath}/generated.csv`,
+    )
+    console.log(`uploaded csv to ${filePath}/generated.csv`)
+
+    const svgTmpDirPath = `/tmp/${filePath}/svg`
+    const svgS3ZipPath = `${filePath}/generated_svg.zip`
+    await fsUtils.fsMkdirOverwriteSync(svgTmpDirPath)
```
[the remark this answers is not in the document]
Comment (gweiying) [reply src/server/serverless/bulk-qrcode-generation/index.js:24]: hmm yea the nesting is a side effect from how the files are saved in the lambda - `/tmp/<folder_id>/png/generated_png.zip`. I'm not quite sure how to change this yet.. The `tmp` and `tmp 2` is a MacBook thing that doesn't allow you to have two folders of the same name if I'm not wrong!

[the remark this answers is not in the document]
Comment (halfwhole) [reply src/server/serverless/bulk-qrcode-generation/index.js:24]: > The `tmp` and `tmp 2` is a MacBook thing that doesn't allow you to have two folders of the same name if I'm not wrong!

right 😅 my bad

File: src/server/serverless/bulk-qrcode-generation/s3.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    archive.pipe(writeStream)
+
+    // where to archive from
+    archive.directory(systemPath)
+    archive.finalize()
+  })
+}
+
+module.exports.uploadToS3 = uploadToS3
+module.exports.streamToS3 = streamToS3
```
[the remark this answers is not in the document]
Comment (gweiying) [reply src/server/serverless/bulk-qrcode-generation/s3.js]: [66c0ca9](https://github.com/opengovsg/GoGovSG/pull/2033/commits/66c0ca981d37627507174b73bcf22950b3a152e1)

File: src/server/serverless/bulk-qrcode-generation/qrCode.js [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+  gov: '#384A51',
+  edu: '#2B2E4A',
+  health: '#472F40',
+}
+const dark = darkColorMap[ASSET_VARIANT]
+const domainMap = {
+  gov: 'go.gov.sg',
+  edu: 'for.edu.sg',
+  health: 'for.sg',
+}
```
[the remark this answers is not in the document]
Comment (gweiying) [reply src/server/serverless/bulk-qrcode-generation/qrCode.js]: [43567dd](https://github.com/opengovsg/GoGovSG/pull/2033/commits/43567dd94d32691fe00285a44451a9736f459faf)

File: src/server/serverless/bulk-qrcode-generation/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (gweiying) [reply src/server/serverless/bulk-qrcode-generation/index.js:24]: ah haha thanks for asking about it I went to read the archiver library more closely and I realised that we can actually set the stream dir to stream into the root - [615c051](https://github.com/opengovsg/GoGovSG/pull/2033/commits/615c0516ad36ef954f9cbce4796256e6ee47a31d)

[the remark this answers is not in the document]
Comment (gweiying) [reply src/server/serverless/bulk-qrcode-generation/index.js:24]: Screenshot 2022-11-15 at 12 03 44
Downloaded file structure after change

File: src/server/serverless/bulk-qrcode-generation/qrCode.js [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+const IMAGE_WIDTH = 1000
+const QR_CODE_DIMENSIONS = 800
+const MARGIN_VERTICAL = 85
+const FONT_SIZE = 32
+const LINE_HEIGHT = 1.35
+
+const ImageFormat = {
+  SVG: 'image/svg+xml',
+  PNG: 'image/png',
+  JPEG: 'image/jpeg',
```
[the remark this answers is not in the document]
Comment (gweiying) [reply src/server/serverless/bulk-qrcode-generation/qrCode.js:17]: Hm.. good point. I think it's ok to leave it for now for consistency, and clean up both server and serverless together if we confirm it's unused?

**Comment by gweiying** [comment]
> LGTM, tested it out and the downloaded QR codes look good, really appreciate the detailed notes esp on testing :D unfortunately I don't know enough of the bigger picture (what happens before SQS -> lambda, and what happens after lambda -> SNS), but as this stands by itself it looks great!
> 
> some other thoughts:
> 
> * what access controls would we like for the S3 bucket for QR codes?
> * would it be good to migrate our lambda functions from javascript -> typescript eventually?
> * will we be using `sqs.js` in the future?
> * seems to be some code duplication w the main codebase for qr code generation, but indeed there doesn't seem to be much we can do about it :')

> * what access controls would we like for the S3 bucket for QR codes? - thinking to leave it as public and having the frontend return the download link directly from the S3 bucket for now (or perhaps with a cloud front distribution infront of it) - this would be the simplest method otherwise we would have to generate a signed url via the server to allow for downloads
> * would it be good to migrate our lambda functions from javascript -> typescript eventually? - YES - would love to, just left it as js for now. if we migrate to ts, we would probably have to migrate all our functions together so that the deployment step can be shared, which would mean quite a bit of refactoring at once. 
> * will we be using `sqs.js` in the future? -> seems no but no harm leaving for now
> * seems to be some code duplication w the main codebase for qr code generation, but indeed there doesn't seem to be much we can do about it :') -> it's the easiest way out but maybe we can look into packaging from a shared folder - not 100% sure it's worth the effort yet

File: src/server/serverless/bulk-qrcode-generation/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (gweiying) [reply src/server/serverless/bulk-qrcode-generation/index.js:24]: @halfwhole, thoughts?

[the remark this answers is not in the document]
Comment (halfwhole) [reply src/server/serverless/bulk-qrcode-generation/index.js:24]: sounds good to me!

File: src/server/serverless/bulk-qrcode-generation/qrCode.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (halfwhole) [reply src/server/serverless/bulk-qrcode-generation/qrCode.js:17]: yup sounds good too

**Review by halfwhole — APPROVED** [review]

**Pull request merged**